src/main/java/com/fasterxml/jackson/datatype/jdk8/OptionalSerializer.java

```diff
@@ -315,6 +315,7 @@ public void acceptJsonFormatVisitor(JsonFormatVisitorWrapper visitor, JavaType t
             }
             _dynamicSerializers = _dynamicSerializers.newWith(type, ser);
         }
+        ser = (JsonSerializer<Object>)provider.handlePrimaryContextualization(ser, _property);
```

Comment:
I can see that lack of contextualization may be a problem, just 2 questions:

1. Shouldn't this be done in `_findSerializer` (since it gets `_property` passed)
2. If not there, shouldn't it be before adding, to avoid having to contextualize more than once per type -- contextualization can be very expensive making calls an order of magnitude slower